src/lib/comms/include/sol-mqtt.h

```diff
@@ -0,0 +1,393 @@
+ *
+ * @return New mqtt object on sucess, NULL otherwise
+ */
+struct sol_mqtt *sol_mqtt_connect(const char *host, int port, const struct sol_mqtt_config *config, void *data);
+
+/**
+ * @brief Reestablish the connection to the MQTT broker
+ *
+ * @param mqtt MQTT Object
+ *
+ * @return 0 on success, -EINVAL otherwise
+ */
+int sol_mqtt_reconnect(struct sol_mqtt *mqtt);
+
+/**
+ * @brief Disconnect from MQTT Broker
+ *
+ * @param mqtt MQTT Object;
+ *
+ * @see sol_mqtt_connect
+ *
+ * Terminate the connection to the broker and free the resources
+ * associated to the mqtt object.
+ */
+void sol_mqtt_disconnect(struct sol_mqtt *mqtt);
+
+/**
+ * @brief Get connection status
+ *
+ * @param MQTT Object
+ *
+ * @see sol_mqtt_conn_status
+ *
+ * @return sol_mqtt_conn_status code
+ */
+int sol_mqtt_get_connection_status(struct sol_mqtt *mqtt);
+
+/**
+ * @brief Send the Broker a message to be published in a given topic
+ *
+ * @param mqtt MQTT Object
+ * @param message Message to be published. The memory associated to this object should be handled by the caller.
+ *
+ * @return 0 on success, -EINVAL otherwise
+ */
+int sol_mqtt_publish(struct sol_mqtt *mqtt, struct sol_mqtt_message *message);
+
+/**
+ * @brief Ask the Broker to be subscribed to a given topic
+ *
+ * @params mqtt MQTT Object
+ * @params topic Topic to subscribe
+ * @param topic_len Topic length in bytes
+ * @param qos MQTT QOS that should be used by the subscribe message
+ *
+ * @return 0 on success, -EINVAL otherwise
+ */
+int sol_mqtt_subscribe(struct sol_mqtt *mqtt, char *topic, size_t topic_len, sol_mqtt_qos qos);
+
+/**
+ * @brief Create a new MQTT message
+ *
+ * @param topic Topic where the message will be published
+ * @param topic_len Length in bytes of the topic
+ * @param payload Payload of the message
+ * @param payload_len Length in bytes of the payload
+ * @param qos MQTT QOS that should be used when sending the message
+ * @param retain If the message should be retained by the broker
+ *
+ * @return New message object on success, NULL otherwise
+ */
+struct sol_mqtt_message *sol_mqtt_message_new(char *topic, size_t topic_len, void *payload,
+        size_t payload_len, sol_mqtt_qos, bool retain);
+
+/**
+ * @brief Free the memory associated to a given message
+ *
+ * @param message MQTT message to be copied
+ *
+ * @return Copy of message on success, NULL otherwise
+ */
+struct sol_mqtt_message *sol_mqtt_message_copy(struct sol_mqtt_message *message);
+
+/**
+ * @brief Free the memory associated to a given message
+ *
+ * @param message MQTT message to be freed
+ */
+void sol_mqtt_message_free(struct sol_mqtt_message *message);
+
+/**
+ * @brief Get the topic of a given MQTT message
+ *
+ * @param message MQTT message to get the topic from
+ * @param buf Buffer to write the topic to
+ * @param size Size in bytes of buf
+ *
+ * @return Number of bytes written to buf
+ */
+size_t sol_mqtt_get_topic(const struct sol_mqtt_message *message, char *buf, size_t size);
```

Comment:
that's the case with most APIs, particularly those in Soletta -- so there is no issue, but it would be good to document that the memory is valid until the caller returns, the main loop runs and other conditions that may change it.

It's just a better pattern to return, eventually you just want to print and exit. Also, you can't know the size to properly allocate a buffer.. returning a "const char*" will let user do strdup(), strlen() and friends.